core-bundle/contao/forms/FormAltcha.php

```diff
@@ -0,0 +1,113 @@
+<?php
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao;
+
+use Contao\CoreBundle\Controller\AltchaController;
+use Contao\CoreBundle\String\HtmlAttributes;
+
+class FormAltcha extends Widget
+{
+	public HtmlAttributes $altchaAttributes;
+
+	/**
+	 * @var boolean
+	 */
+	protected $blnForAttribute = true;
+
+	/**
+	 * @var string
+	 */
+	protected $strTemplate = 'form_altcha';
+
+	/**
+	 * @var string
+	 */
+	protected $prefix = 'widget widget-altcha';
+
+	/**
+	 * Use the raw request data.
+	 *
+	 * @param array $arrAttributes An optional attributes array
+	 */
+	public function __construct($arrAttributes=null)
+	{
+		parent::__construct($arrAttributes);
+
+		$this->useRawRequestData = true;
+	}
+
+	/**
+	 * Generate the widget and return it as string.
+	 *
+	 * @return string The widget markup
+	 */
+	public function generate(): string
+	{
+		return sprintf('<altcha-widget%s></altcha-widget>', $this->getAltchaAttributes());
+	}
+
+	/**
+	 * Parse the template file and return it as string.
+	 *
+	 * @param array $arrAttributes An optional attributes array
+	 *
+	 * @return string The template markup
+	 */
+	public function parse($arrAttributes = null): string
+	{
+		$request = $this->getContainer()->get('request_stack')->getCurrentRequest();
+
+		if ($request && $this->getContainer()->get('contao.routing.scope_matcher')->isBackendRequest($request))
+		{
+			$objTemplate = new BackendTemplate('be_wildcard');
+			$objTemplate->title = $this->label;
+
+			return $objTemplate->parse();
+		}
+
+		$this->altchaAttributes = new HtmlAttributes();
+		$this->altchaAttributes->set('name', $this->name);
+		$this->altchaAttributes->set('challengeurl', $this->getContainer()->get('router')->generate(AltchaController::class));
+		$this->altchaAttributes->set('strings', $this->getLocalization());
```

Comment:
```suggestion
		$this->altchaAttributes->set('strings', $this->getLocalization());
		$this->altchaAttributes->set('maxnumber',$this->getContainer()->get('contao.altcha')->getRangeMax());
```

We need to set the `maxnumber` here (and add the getRangeMax to the service), because the widget splits the workload across the full range so it would waste a lot of energy for calculation numbers that are always too high.

Reply:
> But doesn‘t this lower the security level

It does not IMO. The attacker would start to calculate the hashes beginning from zero anyways and multithreading is not important for spam scripts (and even if it was they could split the number ranges differently so that they do not need to know the maxnumber).